Cython/Compiler/Cache.py

```diff
@@ -0,0 +1,190 @@
+        elif cache:
+            self.path = cache
+            self.enabled = True
+        else:
+            self.path = None
+            self.enabled = False
+        self.cache_size = cache_size if cache_size is not None else 1024 * 1024 * 100
+        if self.enabled:
+            if not os.path.exists(self.path):
+                os.makedirs(self.path)
+
+    def transitive_fingerprint(self, filename, dependencies, compilation_options, language, py_limited_api=False, np_pythran=False):
+        r"""
+        Return a fingerprint of a cython file that is about to be cythonized.
+
+        Fingerprints are looked up in future compilations. If the fingerprint
+        is found, the cythonization can be skipped. The fingerprint must
+        incorporate everything that has an influence on the generated code.
+        """
+        try:
+            m = hashlib.sha1(__version__.encode('UTF-8'))
+            m.update(file_hash(filename).encode('UTF-8'))
+            for x in sorted(dependencies.all_dependencies(filename)):
+                if os.path.splitext(x)[1] not in ('.c', '.cpp', '.h'):
+                    m.update(file_hash(x).encode('UTF-8'))
+            # Include the module attributes that change the compilation result
+            # in the fingerprint. We do not iterate over module.__dict__ and
+            # include almost everything here as users might extend Extension
+            # with arbitrary (random) attributes that would lead to cache
+            # misses.
+            m.update(
+                str((language, py_limited_api,np_pythran)).encode('UTF-8')
+            )
+            m.update(compilation_options.get_fingerprint().encode('UTF-8'))
+            return m.hexdigest()
+        except OSError:
+            return None
+
+    def fingerprint_file_base(self, cfile, fingerprint):
+        return join_path(self.path, "%s-%s" % (os.path.basename(cfile), fingerprint))
+
+    def gz_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + gzip_ext
+
+
+    def zip_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + '.zip'
+
+    def lookup_cache(self, c_file, fingerprint):
+        # Cython-generated c files are highly compressible.
+        # (E.g. a compression ratio of about 10 for Sage).
+        if not os.path.exists(self.path):
+            safe_makedirs(self.path)
+        gz_fingerprint_file = self.gz_fingerprint_file(c_file, fingerprint)
+        zip_fingerprint_file = self.zip_fingerprint_file(c_file, fingerprint)
+        if os.path.exists(gz_fingerprint_file) or os.path.exists(zip_fingerprint_file):
+            if os.path.exists(gz_fingerprint_file):
+                os.utime(gz_fingerprint_file, None)
+                with contextlib.closing(gzip_open(gz_fingerprint_file, 'rb')) as g:
+                    with contextlib.closing(open(c_file, 'wb')) as f:
```

Comment:
```suggestion
                with gzip_open(gz_fingerprint_file, 'rb') as g:
                    with open(c_file, 'wb') as f:
```

I know this is just copied from elsewhere, but I'm just taking the opportunity to clean this up while we're at it.

I think these types have worked in a `with` statement directly for a while

Reply:
Feel free to reject these though if you want to be really strict about just copying. I can come back and do them later.

Reply:
No I like it. I added them to the PR.